File: 232-implement-queue-using-stacks/232-implement-queue-using-stacks.go

// 232. Implement Queue using Stacks
// time: avg/amortized O(1) worst O(n) for every empty pop/peek 
// space: O(n) store 

type MyQueue struct {
    tailStack []int
    headStack []int
}


func Constructor() MyQueue {
    return MyQueue{make([]int,0), make([]int, 0)}
}


func (this *MyQueue) Push(x int)  {
    this.tailStack = append(this.tailStack, x)
}



func (this *MyQueue) Pop() int {
    item := this.Peek()
    n := len(this.headStack)
    this.headStack = this.headStack[:n-1]
    return item
}


func (this *MyQueue) Peek() int {
    if len(this.headStack) == 0 {
        n := len(this.tailStack)
        for n != 0 {
            v := this.tailStack[n-1]
            this.tailStack = this.tailStack[:n-1] 
            this.headStack = append(this.headStack, v)
            n--
        }
    }
    idx := len(this.headStack)-1
    return this.headStack[idx]
}


func (this *MyQueue) Empty() bool {
    return len(this.tailStack) == 0 && len(this.headStack) == 0
}


/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Peek();
 * param_4 := obj.Empty();
 */

File: 232-implement-queue-using-stacks/232-implement-queue-using-stacks.js

var MyQueue = function() {
    this.inStack = []
    this.outStack = []
};

/** 
 * @param {number} x
 * @return {void}
 */
MyQueue.prototype.push = function(x) {
    this.inStack.push(x)
};

/**
 * @return {number}
 */
MyQueue.prototype.pop = function() {
    this.peek()
    n = this.outStack.length
    v = this.outStack[n-1]
    this.outStack.pop()
    return v
};

/**
 * @return {number}
 */
MyQueue.prototype.peek = function() {
    if (this.outStack.length == 0) {
        while (this.inStack.length != 0) {
            v = this.inStack.pop()
            this.outStack.push(v)
        }
    }
    n = this.outStack.length
    return this.outStack[n-1]
};

/**
 * @return {boolean}
 */
MyQueue.prototype.empty = function() {
    return this.inStack.length == 0 && this.outStack.length == 0
};




/** 
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */